pandas-stubs/core/frame.pyi

```diff
@@ -188,7 +188,7 @@ class DataFrame(NDFrame, OpsMixin):
         data: ListLikeU
         | DataFrame
         | dict[Any, Any]
-        | Iterable[tuple[Hashable, ListLikeU]]
+        | Iterable[ListLikeU | tuple[Hashable, ListLikeU]]
```

Comment:
Currently (on main), data is too wide and too narrow: it allows `str` but doesn't allow generators that do not yield `tuple[Hashable, ListLikeU]`. To be clear, some Generators are allowed as they are just Iterables.

If we want to be too wide, I would recommend:

```py
        data: DataFrame
        # column-wise (can enforce Hashable column 'name')
        | Mapping[Hashable, Iterable]
        # row-wise (cannot enforce Hashable column 'name')
        | Iterable[Iterable]
        # None
        | None = ...,
```
and add a test-case:
```py
from typing import Iterable

import pandas as pd

data: Iterable[Iterable] = iter(
    [
        iter([1, 2.0, "something", pd.Timestamp("1/1/2022")]),
    ]
)
pd.DataFrame(data=data)
```
and a similar one for column-wise.

Reply:
> If we want to be too wide

I have a strong preference for being as narrow as possible (but not too narrow).  Prefer the whack-a-mole approach of adding types that are acceptable.

Reply:
Wouldn't mind a follow-up PR to separate row/column-wise initialization (and obviously also to make the annotations more narrow: `Iterable[Sequence]` can be a string - but that might be more tricky). The example from the issue was row-wise, but the added annotations are for what was previously meant for the column-wise case (with the current annotations `tuple[Hashable, ListLikeU]` can be removed as a `tuple` is a `Sequence` which is part of `ListLikeU`).